src/util/renderers/visual-editor.renderer.ts

```diff
@@ -15,37 +14,37 @@ export class VisualEditorRenderer extends Renderer {
     super()
   }
 
-  public async render(renderOpts: RenderOpts): Promise<any> {
-    const { data, isMainPage, articleId, articleDetail } = renderOpts
+  private getDisplayTitleFromVisualEditor(strippedTitle: string, articleId: string) {
+    return strippedTitle || articleId.replace('_', ' ')
+  }
+
+  private async getHTML(renderOpts: RenderOpts): Promise<any> {
+    const { data, articleId, articleDetail, isMainPage } = renderOpts
 
     if (!data) {
       throw new Error(`Cannot render [${data}] into an article`)
     }
 
+    let html: string
+    let displayTitle: string
     let strippedTitle: string
-    const result = []
+
     if (data.visualeditor) {
       // Testing if article has been deleted between fetching list and downloading content.
       if (data.visualeditor.oldid === 0) {
         logger.error(DELETED_ARTICLE_ERROR)
         throw new Error(DELETED_ARTICLE_ERROR)
       }
-      const dataHtml = isMainPage ? data.visualeditor.content : this.injectHeader(data.visualeditor.content, articleDetail)
-      strippedTitle = getStrippedTitleFromHtml(dataHtml)
-      result.push({
-        articleId,
-        displayTitle: strippedTitle || articleId.replace('_', ' '),
-        html: dataHtml,
-      })
-      return result
+      html = isMainPage ? data.visualeditor.content : super.injectH1TitleToHtml(data.visualeditor.content, articleDetail)
+      strippedTitle = getStrippedTitleFromHtml(html)
+      displayTitle = this.getDisplayTitleFromVisualEditor(strippedTitle, articleId)
```

Comment:
I don't know all the details around these two methods, but no way to make simpler? Maybe in one pass?

Reply:
We should here rename to talk about the `H1Title`, so all the function related to this "title" have a similar name.

Reply:
I removed `this.getDisplayTitleFromVisualEditor()` in favor of the explicit title parsing like `displayTitle = strippedTitle || articleId.replace('_', ' ')`